README.md
# Cutespirit Discord Bot
## Install and Run
```sh
git clone https://github.com/Cutespirit-Team/CutespiritDiscordBot

# install requirements
cd ./CutespiritDiscordBot/
pip install -r requirments.txt

# execute
cd ./src
python -m ctbot
```

## Config Sample
[Configuration Example](config-example/)
bot.int
```ini
[bot]
# First run you need to change token to your bot token.
# you can create a bot from 
# https://discord.com/developers/applications
token = MTa4MjS3ASI ....
owner = 196207520585351180

[command]
enable_slash = true
general_prefix = !

[slash]
# regist all slash command for specific guilds.
# accept list of guilds seprate by comma.
all = 875896592052420618,196207701661843456

# regist specific slash command for specific guilds.
# <class name without 'Slash' prefix>.<method> = <guild_ids> 
example.ping = 866199579014987816,96230004047740928

# specific setting for guild
[guild_866199579014987816]
# only response message in specific channels (not work in slash commands)
responsible_channels = 866213047016882206,866977922929917972
```
## 關於我們
[Team Website](https://www.cutespirit.org) <br>
[Team Facebook](https://fb.cutespirit.org) <br>
[Team Discord](https://discord.cutespirit.org)<br>
[Team Telegram](https://telegram.cutespirit.org)<br>
[Cutespirit-SHOP](https://shop.cutespirit.org)<br>
[Cutespirit Discord Bot](https://dcbot.cutespirit.org)<br>

# 趨勢
[![Stargazers over time](https://starchart.cc/Cutespirit-Team/CutespiritDiscordBot.svg)](https://starchart.cc/Cutespirit-Team/CutespiritDiscordBot)


Licence:© Cutespirit 2022 All right reversed 此程式除了「關於」頁面不可重製及發布之外，其餘頁面及功能可進行重製發布。


changelog.md
| 版本 | 更新日期 | 更新內容 |
|-----|-----|-----|
| v0.1 | 2021/07/19 | 初始版本(由TershiCloud Telegram Bot修改) |
| v0.2 | 2021/07/19 | 增加/status指令，可以隨時更改Discord狀態，刪除lang語言功能。 |
| v0.3 | 2021/07/20 | 增加/kick /ban /unban指令，可以隨時踢人，封鎖，解封人。 |
| v0.4 | 2021/07/20 | 將/time --help顯示/week問題修復。 |
| v0.5 | 2021/07/20 | 完善各--help幫助內容。 |
| v0.6 | 2021/07/20 | 修復Bug。 |
| v0.7 | 2021/07/20 | 並加入YouTube之/join /play /pause /resume /stop /leave功能。 |
| v0.8 | 2021/07/21 | 完善沒有加參數無法顯示的錯誤。 |
| v0.9 | 2021/07/22 | 修復/play二次加入連結顯示沒有播放的錯誤。 |
| v0.9.1 | 2021/07/22 | 刪除/showweb指令。 |
| v0.9.2 | 2021/07/22 | 增加一般群組無法執行/命令。 |
| v1.0 | 2021/07/22 | 將所有help與傳送字樣中文化，並且將message一部分導出。 |
| v2.0 | 2021/08/14 | Completely rewrite |
| v2.1 | 2021/08/28 | Update Slash Library to 2.3.2|
| v2.2 | 2021/08/29 | Update Help Command |
| v2.3 | 2021/08/29 | Featrue: add weather infomations|
| v2.4 | 2021/08/29 | Feature: add Wx, PoP, MinT|
| v2.5 | 2021/08/29 | Fix Bug: input 台北 auto convert to 臺北市|
| v2.5.1 | 2021/08/29 | Change long weatherElement Code to be simple |
| v2.5.2 | 2021/08/30 | Change Weather display mode and add option | 
| v2.6 | 2021/12/30 | Add Feature: Time Remain of this year |
| v2.6.1 | 2021/12/31 | Add Feature: Time Remain in roc or vids |
| v2.6.2 | 2022/01/02 | Add Feature: Show Hardware information |
| v2.6.3 | 2022/01/02 | Fix Bug: add hardware Description and help |
| v2.6.4 | 2022/01/02 | Add Feature: Run Speedtest |
| v2.6.5 | 2022/01/09 | Add Feature: Time Exam Left |
| v2.6.6 | 2022/01/10 | Add Feature: Github Info |
| v2.6.7 | 2022/01/10 | Add Feature: Server State |
| v2.6.8 | 2022/01/11 | Update github_info Not Found |
| v2.6.9 | 2022/01/18 | Add Feature: increase github_visitor |
| v2.7.0 | 2022/01/21 | Add Feature: check if have permission or not |
| v2.7.1 | 2022/01/25 | Fix bug: change weather token to version.py |
| v2.8 | 2022/01/27 | Feature: Add Cutespirit-SHOP Place Order | Feature: Add Cutespirit-SHOP Check Order | Feature: Add About dcbot, author, team | Fix bug: Add version.py more configurations | Update version to about |
| v2.9 | 2022/05/05 | Feature: Add words forbidden |
| v3.0 | 2022/05/05 | Feature: Add Time Config| Feature: Add words forbidden Config | Feature: Add user customize time left | Feature: Remove time exam left |
| v3.1 | 2022/05/07 | Feature: Add Member_Join_Notification | Feature: Add Member_Left_Notificaion | Fix Bug: Typo at words forbidden | Add Feature: Member_Join_Notification Embed | Featrue: Add Member Config | Feature: Add create invitation link |
| v3.2 | 2022/05/08 | Add Fearute: Service Ticket | Change Feature: help text to embed |
| v3.3 | 2022/05/08 | Add Feature: guild Config for Specific Guild |
| v3.4 | 2022/05/12 | Feature: Add TaiwanStock | Feature: Add Help Q and A | Feature: Add service ticket | Fix Bug: weather not respond |
| v3.5 | 2022/05/12 | Fix Bug: utils.py json import error | Fix Bug: remove development mode | Feature: Add stauts type |
| v3.6 | 2022/05/14 | Feature: Add Reaction Role | Fix bug: time zone | Feature: Add spam words detect | Feature: Add specific channel id when specific reaction roles | Feature: Remove notification when opening a ticket | Add Feature: Give guest role when a member join | Files: Add configuration example |
| v3.7 | 2022/05/14 | Fix Bug: member join error | Feature: when ticket closed, move it to category | Fix Bug: Taiwan stock output error |
| v4.0 | 2022/06/25 | Feature: Add normal role of clicking reaction | Feature: Add reload command | Feature: Remove config recreating | Feature: Add reload status notification | Feature: Add sendmsg permission | Feature: Integrate member_count to server_state | Feature: Add YouTube channel info | Feature: Add YouTube video info |
| v4.0.1 | 2022/08/29 | Fix Bug: member join leave will not send noti |

src/ctbot/config.py
import os
import configparser
import typing

'''
config sample
[bot]
token = MTa4MjS3ASI ....
owner = 196207520585351180

[command]
enable_slash = true
general_prefix = !

[slash]
all = <guild_id>
<class without 'Slash' prefix>.<method> = <guild_id>
example.ping = 866199579014987816

# 個別 guilds 設定
[guild_866199579014987816]

# 只響應指定頻道的訊息以逗號分隔, 如果為空響應所有頻道
responsible_channels = 866213047016882206,866977922929917972

'''

def to_ints(ids: typing.List[str]) -> typing.List[int]:
    return set(map(int, filter(str.isdigit, ids)))

def parse_bool(s: str):
    return s.lower() in ['on', '1', 'true']

class BotConfig:
    config_dir = os.path.dirname(os.path.realpath(__file__))

    def __init__(self, filename: str='config.ini'):
        self.filename = filename
        self.path = os.path.realpath(os.path.join(BotConfig.config_dir, filename))
        self.guilds = {}
        self.load()

    def save(self):
        with open(self.path, 'w') as configfile:
            self._config.write(configfile)

    def load(self):
        parser = configparser.ConfigParser()
        if os.path.exists(self.path):
            parser.read([self.path])
        else:
            BotConfig.create_default(parser, save_to=self.path)
        self._config = parser
        self._parse_guilds_section()
    
    def _parse_guilds_section(self):
        
        type_mapping = {
            'responsible_channels': {'type': list},
            '__default__': {'type': str}
        }

        sections = filter(lambda section: section.startswith('guild_'), self._config.sections())
        guilds = {}
        for section in sections:
            guild_id = section.split('_')[1]
            guilds.setdefault(guild_id, {})
            for key, value in self._config.items(section):
                t = type_mapping.get(key, type_mapping.get('__default__'))
                if t.get('type') == list:
                    value = to_ints(value.split(','))
                guilds[guild_id].setdefault(key, value)
        self.guilds = guilds
    
    @classmethod
    def create_default(cls, parser: configparser.ConfigParser, save_to=None):
        parser.read_dict({
            'bot': {
                'token': '',
                'owner': '',
            },
            'command': {
                'enable_slash': 'false',
                # 'enable_general': 'true',
                'general_prefix': '!',
            },
            'slash': {
                'all': ''
            }
        })

        if save_to:
            with open(save_to, 'w') as f:
                parser.write(f)

    def get_token(self):
        return self._config.get('bot', 'token', fallback='')

    def get_owner(self):
        return self._config.get('bot', 'owner', fallback='')

    def get_enable_slash(self):
        return parse_bool(self._config.get('command', 'enable_slash', fallback='false'))

    def get_enable_general(self):
        return parse_bool(self._config.get('command', 'enable_general', fallback='false'))

    def get_general_prefix(self):
        return self._config.get('command', 'general_prefix', fallback='!')
     
    def get_responsible_channels(self, guild: int) -> typing.List[int]:
        guild = self.guilds.get(str(guild), {})
        channel_ids = guild.get('responsible_channels', set())
        return channel_ids

    def get_slash_command_guilds_id(self, module: str, method: str) -> typing.List[int]:
        all_guilds_id = self._config.get('slash', 'all', fallback='').split(',')
        method_guilds_id = self._config.get('slash', f'{module}.{method}', fallback='').split(',')
        return to_ints(all_guilds_id + method_guilds_id)

def _generate_getter_function(section: str, option: str, _type: type):
    def getter_wrapper(self):
        value = self._config.get(section, option, fallback='')
        if _type == list:
            value = value.split(',')
        return value

    def getter_guild_wrapper(self, guild_id: str):
        value = self._config.get(guild_id, option, fallback='')
        if _type == list:
            value = value.split(',')
        return value

    return getter_guild_wrapper if section == 'guild' else getter_wrapper
    
'''
    {'name': 'responsible_channels', 'section': 'guild'}
    setattr(CTConfig, ...)
'''

# config = CTBotConfig('config.ini')
# print(config.get_slash_command_guilds_id('activity', 'change'))
